src/messages/manifestjson.js

```diff
@@ -676,3 +717,53 @@ export const EXTENSION_ID_REQUIRED = {
   ),
   file: MANIFEST_JSON,
 };
+
+export const PRIVILEGED_FEATURES_REQUIRED = 'PRIVILEGED_FEATURES_REQUIRED';
+export function privilegedFeaturesRequired(error) {
+  const messageTmpl = i18n._(
+    oneLine`%(instancePath)s: Privileged extensions should declare privileged permissions.`
+  );
+
+  const message = i18n.sprintf(messageTmpl, {
+    instancePath: error.instancePath,
+  });
+
+  return {
+    code: PRIVILEGED_FEATURES_REQUIRED,
+    message,
+    description: i18n._(oneLine`
+      This extension does not declare any privileged permission. It does not need to be signed with the privileged certificate.
+      Please upload it directly to "https://addons.mozilla.org/".
```

Comment:
@willdurand while I was going through the last round of suggested edits I looked back to this description and I'm wondering if this may sounds like something the linter is 100% sure of and then pushing privileged extensions developers to don't even double-check if the addon was actually expected to be privileged.

Besides the couple of corner case that the linter can't detect at the moment, there is a chance that the developer may have missed to specify a permission or mistyped a property which would prevent the addon to be detected as requiring a privileged signature (but may be only breaking a secondary feature).

I'm also thinking that we could also detect if the manifest is including a restricted match pattern (which require also the "mozillaAddons" permission to work as expected) and use also that as an hint that the add-on is expected to be privileged. 

wdyt? 

do we want to leave that as a separate followup to this PR?

Reply:
Yeah, I think that's why we'll need to enforce `"mozillaAddons"` for privileged extensions in Firefox in the future. That will be the best way for developers to think about that and know whether their extension is privileged or not.

In the meantime, we can probably detect restricted match patterns indeed. I would think that a separate PR could be easier to review but up to you really.

Reply:
k, I'll create a separate PR for the restricted match pattern detection.

For the Firefox side, we could consider to introduce sooner a warning logged on missing "mozillaAddons" permission in extensions detected as privileged, that wouldn't break backward compatibility and may provide more visible hints to the extension developers hitting it.